cassiopeia/data/CassiopeiaTree.py

```diff
@@ -2046,6 +2061,315 @@ def get_unmutated_characters_along_edge(
+                            and "*" not in c
+                        ):
+                            child_possible_states.append(c)
+                    # '*' represents all non-missing states, and is a
+                    # possible state when all children are missing, as any
+                    # state could have occurred before if all missing states
+                    # occurred independently.
+                    if child_possible_states == []:
+                        possible_states[char] = [
+                            "*",
+                            self.missing_state_indicator,
+                        ]
+                    else:
+                        possible_states[char] = list(
+                            set.intersection(*child_possible_states)
+                        )
+                        if 0 not in possible_states[char]:
+                            possible_states[char].append(0)
+
+            likelihoods_at_n = [{} for i in range(self.n_character)]
+
+            # We calculate the likelihood of the states at each character at
+            # the current node according to the recurrence relation
+            for char in range(self.n_character):
+                for s in possible_states[char]:
+                    likelihood_s = 1
+                    for child in self.children(n):
+                        likelihood_for_child = 0
+                        for s_ in likelihoods_at_nodes[child][char]:
+                            likelihood_for_child += (
+                                transition_probability(
+                                    s,
+                                    s_,
+                                    self.get_branch_length(n, child),
+                                    char,
+                                )
+                                * likelihoods_at_nodes[child][char][s_]
+                            )
+                        likelihood_s *= likelihood_for_child
+                    likelihoods_at_n[char][s] = likelihood_s
+
+            likelihoods_at_nodes[n] = likelihoods_at_n
+
+        # If we are not to use the internal state annotations explicitly,
+        # then we impose an implicit root of all 0s if it does not exist
+        if (
+            not use_internal_character_states
+            and len(self.children(self.root)) != 1
+        ):
+            likelihoods_at_implicit_root = [{} for i in range(self.n_character)]
+            for char in range(self.n_character):
+                likelihood_at_root = 0
+                for s_ in likelihoods_at_nodes[self.root][char]:
+                    likelihood_at_root += (
+                        transition_probability(0, s_, 1, char)
+                        * likelihoods_at_nodes[self.root][char][s_]
+                    )
+                likelihoods_at_implicit_root[char][0] = likelihood_at_root
+
+            return np.log(np.prod([i[0] for i in likelihoods_at_implicit_root]))
```

Comment:
why not use `np.sum([np.log(i[0]) for i in likelihoods_at_implicit_root])`? I think this will be more stable.

Also, let's be clear in the description that we're returning the _log_ likelihood.